ast/format.go

```diff
@@ -64,6 +61,113 @@ func (f *formatter) writeComment(comment string) {
 	f.writeRune('\n')
 }
 
+// Logic for handling operator precedence and parenthesis formatting.
+
+const (
+	functionCall = 1
+	member       = 2
+	index        = 3
+
+	// Use offsets for operators and logical operators to ensure they are unique keys
+	// in the map of operators precedence.
+	opOffset  = 100
+	lopOffset = 1000
+)
+
+func getIntForOp(op OperatorKind) int {
+	return int(op) + opOffset
+}
+
+func getIntForLOp(op LogicalOperatorKind) int {
+	return int(op) + lopOffset
+}
+
+func getPrecedence(key int) int {
+	return opPrecedence[key]
+}
+
+func getPrecedenceForOp(op OperatorKind) int {
+	return getPrecedence(getIntForOp(op))
+}
+
+func getPrecedenceForLOp(op LogicalOperatorKind) int {
+	return getPrecedence(getIntForLOp(op))
+}
+
+// this matches the SPEC
+var opPrecedence = map[int]int{
+	functionCall: 1,
+	member:       1,
+	index:        1,
```

Comment:
Do index and member need to be included? See my comments regarding their associated test cases.

Reply:
I absolutely see your point, but I made it for completeness and in order to directly match the SPEC. I also know that those in the tests wouldn't pass type checking, but in the future they could be (say we support `+` for arrays). So I wanted to check if operator precedence was respected even in that cases, also given that there is actually a formally defined precedence of (say) member VS addition in the SPEC.